### PR TITLE
Delay Compensation ASGD

### PR DESCRIPTION
**Pull request by zuowang**
Implement a new sgd proposed in paper "Asynchronous Stochastic Gradient Descent with Delay Compensation for Distributed Deep Learning"

### CODE REVIEW

**Comment by backyes** [comment]
Can you do more performance data, such as convergence. 

Previously we have implemented several  ASGD variation, try to gain better convergence performance, such as staleness AsyncSGD. We did not merge these patch for following reasons:
* Async-sgd is not stable
* Even sync-sgd gains better system performance with weak lock requirement, but the overall convergence performance could not surpass the sync-sgd (mean 'sgd' in PaddlePaddle by default) with general parameter initialisation.

In addition, we have some technology to help improve the sync-sgd synchronisation performance, such as carefully designed tcp stack, low latency, lossless link, fine grained sync skill in reality, latency hidden skill in distributed system. At last, sync-sgd could be better.

So, we hope for more detailed performance data. :-)

**Comment by yingfeng** [comment]
Actually we do not have enough environment to perform the convergence test of this DC-ASGD, it's the remarkable result from the paper that leads to this pull request, as a result, could you please make tests to see whether such improvements from Microsoft would really work?
By the way, this pull request is based on the default assumption from the publication, which means a high memory consumption (with M backup weight matrices while M equals to number of workers).  We will apply another policy later to remove such kind of memory wasting at a cost of double bandwidth when pushing parameters to PS server, as also mentioned in the publication.

**Comment by backyes** [comment]
Thanks you very much. 

You can try this with several docker VMs to verify the convergence. the simple cluster for cluster train could be avaiable.
Try use PaddlePaddle ubuntu docker image..  

Also, we can plan to try this patch later.

**Comment by luotao1** [comment]
感谢您给PaddlePaddle贡献代码。由于Paddle V1/V2版本已不再维护，相关代码也已从develop分支上删除，因此关闭您的PR，欢迎您向Paddle最新版-Fluid贡献代码。
Thanks for contributing to PaddlePaddle! Since V1/V2 will not be maintained anymore, and related codes have been deleted from develop branch as well, we close this PR. Welcome to contribute to Fluid——the latest version of PaddlePaddle.
